src/server.rs

```diff
@@ -273,38 +273,36 @@ fn send_index(
     Ok(())
 }
 
-const HTREE_SEND_PIPELINE_LEN: usize = 5;
-
 fn send_htree(
     repo: &mut repository::Repo,
     tr: &mut htree::TreeReader,
     w: &mut dyn std::io::Write,
 ) -> Result<(), anyhow::Error> {
     let mut storage_engine = repo.storage_engine()?;
-
     loop {
         match tr.current_height() {
             Some(0) => {
-                let mut chunk_address = address::Address::default();
-                let mut pipeline = std::collections::VecDeque::new();
-
-                for ent_slice in tr.pop_level().unwrap().chunks(8 + address::ADDRESS_SZ) {
-                    let address_slice = &ent_slice[8..];
-                    chunk_address.bytes.clone_from_slice(&address_slice);
-                    pipeline.push_back((
-                        chunk_address,
-                        storage_engine.get_chunk_async(&chunk_address),
-                    ));
-
-                    if pipeline.len() >= HTREE_SEND_PIPELINE_LEN {
-                        let (chunk_address, chunk_data) = pipeline.pop_front().unwrap();
-                        write_chunk(w, &chunk_address, &chunk_data.recv()??)?;
-                    }
-                }
+                let address_buf = tr.pop_level().unwrap();
+
+                let mut addresses =
+                    Vec::with_capacity(address_buf.len() / (8 + address::ADDRESS_SZ));
+
+                // XXX Can we avoid this potentially large allocation/copy?
+                // It serves very little purpose and could double ram usage of this function
+                // in a bad case.
+                addresses.extend(
+                    address_buf
+                        .chunks(8 + address::ADDRESS_SZ)
+                        .map(|x| address::Address::from_slice(&x[8..]).unwrap()),
+                );
 
-                while let Some((chunk_address, chunk_data)) = pipeline.pop_front() {
-                    write_chunk(w, &chunk_address, &chunk_data.recv()??)?;
-                }
+                let mut on_chunk =
+                    |addr: &address::Address, data: Vec<u8>| -> Result<(), anyhow::Error> {
+                        write_chunk(w, addr, &data)?;
+                        Ok(())
+                    };
```

Comment:
Shouldn't this be outside the loop since it's not capturing anything inside it?

Reply:
It's capturing w I think.